experimental/tiledb/common/dag/nodes/detail/segmented/function.h

```diff
@@ -0,0 +1,385 @@
+  }
+
+  /** Utility functions for indicating what kind of node and state of the ports
+   * being used.
+   *
+   * @todo Are these used anywhere?  This is an abstraction violation, so we
+   * should try not to use them.
+   * */
+  bool is_function_node() const override {
+    return true;
+  }
+
+  bool is_source_empty() const override {
+    auto mover = this->get_source_mover();
+    return empty_source(mover->state());
+  }
+
+  bool is_sink_full() const override {
+    auto mover = this->get_sink_mover();
+    return full_sink(mover->state());
+  }
+
+  bool is_source_terminating() const override {
+    auto mover = this->get_source_mover();
+    return terminating(mover->state());
+  }
+
+  bool is_sink_terminating() const override {
+    auto mover = this->get_sink_mover();
+    return terminating(mover->state());
+  }
+  bool is_source_terminated() const override {
+    auto mover = this->get_source_mover();
+    return terminated(mover->state());
+  }
+
+  bool is_sink_terminated() const override {
+    auto mover = this->get_sink_mover();
+    return terminated(mover->state());
+  }
+
+  bool is_source_done() const override {
+    auto mover = this->get_source_mover();
+    return done(mover->state());
+  }
+
+  bool is_sink_done() const override {
+    auto mover = this->get_sink_mover();
+    return done(mover->state());
+  }
+
+  bool is_sink_state_empty() const override {
+    auto mover = this->get_sink_mover();
+    return empty_state(mover->state());
+  }
+
+  bool is_sink_state_full() const override {
+    auto mover = this->get_sink_mover();
+    return full_state(mover->state());
+  }
+
+  bool is_source_state_empty() const override {
+    auto mover = this->get_source_mover();
+    return empty_state(mover->state());
+  }
+
+  bool is_source_state_full() const override {
+    auto mover = this->get_source_mover();
+    return full_state(mover->state());
+  }
+  /**
+   * @brief Get the name of the node.
+   *
+   * @return The name of the node.
+   */
+  std::string name() const override {
+    return {"function"};
+  }
+
+  /**
+   * @brief Enable debug output for this node.
+   */
+  void enable_debug() override {
+    node_base_type::enable_debug();
+    if (SinkBase::item_mover_)
+      SinkBase::item_mover_->enable_debug();
+    if (SourceBase::item_mover_)
+      SourceBase::item_mover_->enable_debug();
+  }
+
+ private:
+  auto get_sink_mover() const {
+    return SinkBase::get_mover();
+  }
+
+  auto get_source_mover() const {
+    return SourceBase::get_mover();
+  }
+
+  void dump_node_state() override {
+    auto source_mover = this->get_source_mover();
+    auto sink_mover = this->get_sink_mover();
+    std::cout << this->name() << " Node state: " << str(sink_mover->state())
+              << " -> " << str(source_mover->state()) << std::endl;
+  }
+
+  BlockIn in_thing{};  // @todo We should use the port item_
+  BlockOut out_thing{};
+
+ public:
+  //#pragma clang diagnostic push
```

Comment:
@lums658 I had to comment this out because it isn't recognized by AppleClang 14.